### PR TITLE
Push the smallfile test results into elasticsearch server.

### PR DESCRIPTION
**Pull request by Avilir**
Results push via ripsaw (snafu) and also with some more information
directly by this test.
Also, increas the number of the samples from 1 to 3.

Signed-off-by: Avi liani <alayani@redhat.com>

Also, it will retrieve the previous test results and compare to current result to see if there is a regression / progression or no change.

### CODE REVIEW

File: ocs_ci/ocs/cluster.py [hunk 1]
```diff
@@ -632,6 +632,34 @@ def time_taken_to_complete_rebalance(self, timeout=600):
                 time_taken = time.time() - start_time
                 return (time_taken / 60)
 
+    def get_clustername(self):
+        """
+        Return the name (DNS short name) of the cluster
+
+        Args:
```
Comment (AviadP) [inline ocs_ci/ocs/cluster.py:639]: no need to mention args if there are none

File: ocs_ci/ocs/cluster.py [hunk 2]
```diff
@@ -632,6 +632,34 @@ def time_taken_to_complete_rebalance(self, timeout=600):
...
+            None
+
+        Returns:
+             str: the short DNS name of the cluster
+        """
+        out = run_cmd('oc get route -n openshift-console -o yaml')
```
Comment (AviadP) [inline ocs_ci/ocs/cluster.py:645]: better using ocp.get here

File: ocs_ci/ocs/cluster.py [hunk 3]
```diff
@@ -632,6 +632,34 @@ def time_taken_to_complete_rebalance(self, timeout=600):
...
+        Return the OCS Version
+
+        Args:
+            None
+
+        Returns:
+             str: The version of the OCS
+        """
+        ns = config.ENV_DATA['cluster_namespace']
+        out = run_cmd(f'oc get csv -n {ns} -o yaml')
```
Comment (AviadP) [inline ocs_ci/ocs/cluster.py:659]: same. (use ocp.get)

File: tests/e2e/performance/test_small_file_workload.py [hunk 4]
```diff
@@ -40,31 +42,66 @@ class TestSmallFileWorkload(E2ETest):
     workloads
     """
 
+    def analize_results(self, results, logs):
```
Comment (keesturam) [inline tests/e2e/performance/test_small_file_workload.py:45]: `analyze_results`

File: tests/e2e/performance/test_small_file_workload.py [hunk 5]
```diff
@@ -40,31 +42,66 @@ class TestSmallFileWorkload(E2ETest):
...
                 marks=pytest.mark.polarion_id("OCS-2022"),
             ),
             pytest.param(
-                *[16, 50000, 4, 1, constants.CEPHFILESYSTEM],
+                *[16, 50000, 4, 3, constants.CEPHFILESYSTEM],
                 marks=pytest.mark.polarion_id("OCS-2023"),
             ),
             pytest.param(
-                *[16, 200000, 4, 1, constants.CEPHFILESYSTEM],
+                *[16, 200000, 4, 3, constants.CEPHFILESYSTEM],
```
Comment (keesturam) [inline tests/e2e/performance/test_small_file_workload.py:104]: We want to remove this test. So, let's ignore updating this param.

File: ocs_ci/ocs/cluster.py (hunk 2 above)
Comment (Avilir) [reply ocs_ci/ocs/cluster.py]: i changed it.

File: ocs_ci/ocs/cluster.py (hunk 3 above)
Comment (Avilir) [reply ocs_ci/ocs/cluster.py]: get_clustername

File: ocs_ci/ocs/cluster.py (hunk 1 above)
Comment (Avilir) [reply ocs_ci/ocs/cluster.py]: removed.

File: tests/e2e/performance/test_small_file_workload.py (hunk 4 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: i fixed it.

File: tests/e2e/performance/test_small_file_workload.py (hunk 5 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: i delete this test.

File: ocs_ci/ocs/ocp.py [hunk 6]
```diff
@@ -840,6 +840,50 @@ def get_logs(
...
+        Return the OCP Build Version
+
+        Args:
+            None
+
+        Returns:
+             str: The build version of the OCP
+        """
+        ns = config.ENV_DATA['cluster_namespace']
+        out = run_cmd(f'oc get clusterversion -n {ns} -o yaml')
```
Comment (AviadP) [inline ocs_ci/ocs/ocp.py:854]: plz avoid using run_cmd if there are better option (ocp.get etc.) - this apply for any place you are using run_cmd

File: ocs_ci/ocs/cluster.py [hunk 7]
```diff
@@ -632,6 +632,30 @@ def time_taken_to_complete_rebalance(self, timeout=600):
...
+        """
+        Return the OCS Version
+
+        Returns:
+             str: The version of the OCS
+        """
+        ocp_cluster = ocp.OCP(
+            namespace=config.ENV_DATA['cluster_namespace'],
+            kind='', resource_name='csv')
+        return ocp_cluster.get()['items'][0]['spec']['version']
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:657]: I don't think these function belong in this class (CephCluster) as they are for the OCP cluster. I would place them in ocp.py, as standalone functions, not OCP class methods

File: ocs_ci/ocs/ocp.py [hunk 8]
```diff
@@ -840,6 +840,43 @@ def get_logs(
         )
         return output
 
+    def get_build(self):
+        """
+        Return the OCP Build Version
+
+        Args:
+            None
```
Comment (ebenahar) [inline ocs_ci/ocs/ocp.py:848]: No need. Please fix in all places

File: ocs_ci/ocs/ocp.py [hunk 9]
```diff
@@ -840,6 +840,43 @@ def get_logs(
...
+
+        Returns:
+             str: The build version of the OCP
```
Comment (ebenahar) [inline ocs_ci/ocs/ocp.py:851]: add an empty line below to comply with Google sphinx style - https://www.sphinx-doc.org/en/1.5/ext/example_google.html

Please fix in all plcaes

File: ocs_ci/templates/workloads/smallfile/SmallFile.yaml [hunk 10]
```diff
@@ -8,12 +8,14 @@ metadata:
   name: smallfile-benchmark
   namespace: my-ripsaw
 spec:
-  test_user: homer_simpson
-  clustername: aws-dec26-2019
+  test_user: avi-liani
```
Comment (ebenahar) [inline ocs_ci/templates/workloads/smallfile/SmallFile.yaml:11]: are these changes intentional?

File: tests/e2e/performance/test_small_file_workload.py [hunk 11]
```diff
@@ -40,40 +42,77 @@ class TestSmallFileWorkload(E2ETest):
     workloads
     """
 
+    def analyze_results(self, results, logs):
```
Comment (ebenahar) [inline tests/e2e/performance/test_small_file_workload.py:45]: please add docstring

File: tests/e2e/performance/test_small_file_workload.py [hunk 12]
```diff
@@ -40,40 +42,77 @@ class TestSmallFileWorkload(E2ETest):
...
+                    results['smallfile-res'][op],
+                    'elapsed_time', line.split()[-1].strip())
+            if "files/sec" in line:
+                results['smallfile-res'][op] = add_value(
+                    results['smallfile-res'][op],
+                    'filesPsec', line.split()[-1].strip())
+            if "requested files" in line:
+                results['smallfile-res'][op] = add_value(
+                    results['smallfile-res'][op],
+                    'percenrage', line.split()[0].strip())
```
Comment (ebenahar) [inline tests/e2e/performance/test_small_file_workload.py:77]: I suggest converting this to dict

File: ocs_ci/templates/workloads/smallfile/SmallFile.yaml (hunk 10 above)
Comment (Avilir) [reply ocs_ci/templates/workloads/smallfile/SmallFile.yaml:11]: no particular reason for the change, will change it back.

File: ocs_ci/ocs/cluster.py (hunk 7 above)
Comment (Avilir) [reply ocs_ci/ocs/cluster.py]: moved the out as suggested.

File: ocs_ci/ocs/ocp.py (hunk 8 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: fixed them all.

File: ocs_ci/ocs/ocp.py (hunk 9 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: update in all places

File: tests/e2e/performance/test_small_file_workload.py (hunk 12 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: replaced.

File: tests/e2e/performance/test_small_file_workload.py (hunk 11 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: function replaced

File: ocs_ci/templates/workloads/smallfile/SmallFile.yaml (hunk 10 above)
Comment (Avilir) [reply ocs_ci/templates/workloads/smallfile/SmallFile.yaml]: changed back.

File: tests/e2e/performance/test_small_file_workload.py [hunk 13]
```diff
@@ -5,24 +5,310 @@
...
+        self.all_results = {}
+
+        # make sure we have connection to the elastic search server
+        log.info(f'Connecting to ES {self.server} on port {self.port}')
+        try:
+            self.es = Elasticsearch([{'host': self.server, 'port': self.port}])
+        except ESExp.ConnectionError:
+            log.error('can not connect to ES server {}:{}'.format(
+                self.server, self.port))
+            raise
```
Comment (vasukulkarni) [inline tests/e2e/performance/test_small_file_workload.py:73]: @Avilir since we are going to run this on AWS as well, I think we should ignore here.

**Comment by vasukulkarni** [comment]
Can you schedule a test run in AWS for this test?

File: ocs_ci/ocs/ocp.py [hunk 14]
```diff
@@ -841,6 +841,78 @@ def get_logs(
         return output
 
 
+def get_ocp_provider():
```
Comment (vasukulkarni) [inline ocs_ci/ocs/ocp.py:844]: s /get_ocp_provider/get_provider/

File: ocs_ci/ocs/ocp.py [hunk 15]
```diff
@@ -841,6 +841,78 @@ def get_logs(
...
+
+    Returns:
+         str: The Provider that the OCP is running on
+    """
+
+    ocp_cluster = OCP(kind='', resource_name='nodes')
+    return ocp_cluster.get('nodes')['items'][0]['spec']['providerID'].split(':')[0]
+
+
+def get_ocp_clustername():
```
Comment (vasukulkarni) [inline ocs_ci/ocs/ocp.py:856]: s/get_ocp_clustername/get_clustername/

File: ocs_ci/ocs/ocp.py [hunk 16]
```diff
@@ -841,6 +841,78 @@ def get_logs(
...
+
+    """
+
+    ocp_cluster = OCP(
+        namespace=config.ENV_DATA['cluster_namespace'],
+        kind='', resource_name='csv')
+    return ocp_cluster.get()['items'][0]['spec']['version']
+
+
+def get_ocp_build():
```
Comment (vasukulkarni) [inline ocs_ci/ocs/ocp.py:886]: same

**Comment by keesturam** [comment]
@Avilir Looks great, Would it be possible to attach the log of sample test run?

File: tests/e2e/performance/test_small_file_workload.py (hunk 13 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py:73]: @vasukulkarni This is not possible since the benchmark sending the results to ES - from within the pod, so without that we do not have full results.
we must have ES with public IP so the benchmark can send results to it, or we need to change the benchmark itself to collect and report the results in different way.

File: ocs_ci/ocs/ocp.py (hunk 14 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: done

File: ocs_ci/ocs/ocp.py (hunk 15 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: done

File: ocs_ci/ocs/ocp.py (hunk 16 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: done

**Comment by Avilir** [comment]
> @Avilir Looks great, Would it be possible to attach the log of sample test run?

You can find sample results at : [here](http://pastebin.test.redhat.com/864049)
as you can see the test Failed since in lines 179, 183 and 187 it report on "Error  : Deviation < > IOPS is more the 5% ()"

will try to trigger Jenkins job (which probably will fails too)

File: tests/e2e/performance/test_small_file_workload.py (hunk 13 above)
Comment (vasukulkarni) [reply tests/e2e/performance/test_small_file_workload.py:73]: we will need a mark to skip this test in AWS which can be taken as follow on PR

File: ocs_ci/ocs/ocp.py [hunk 17]
```diff
@@ -841,6 +841,78 @@ def get_logs(
         return output
 
 
+def get_provider():
+    """
+    Return the OCP Provider (Platform)
+
+    Returns:
+         str: The Provider that the OCP is running on
```
Comment (ebenahar) [inline ocs_ci/ocs/ocp.py:849]: add an empty line below

File: ocs_ci/ocs/ocp.py [hunk 18]
```diff
@@ -841,6 +841,78 @@ def get_logs(
         return output
 
 
+def get_provider():
```
Comment (ebenahar) [inline ocs_ci/ocs/ocp.py:844]: shouldn't this function be in `ocs_ci/ocs/node.py`

File: tests/e2e/performance/test_small_file_workload.py [hunk 19]
```diff
@@ -5,24 +5,316 @@
...
+        }
+
+        # Calculate the number of records for the test
+        self.records = self.results['clients'] * self.results['threads']
+        self.records *= self.results['samples']
+        self.records *= len(self.results['operations'])
+
+    def add_key(self, key, value):
+        """
+        Adding key and value the this object results dict
```
Comment (ebenahar) [inline tests/e2e/performance/test_small_file_workload.py:93]: sentence needs rephrasing

File: tests/e2e/performance/test_small_file_workload.py [hunk 20]
```diff
@@ -5,24 +5,316 @@
...
+
+        Args:
+            key (str): String which will be the key for the value
+            value (*): value to add, can be any kind of data type
```
Comment (ebenahar) [inline tests/e2e/performance/test_small_file_workload.py:97]: what is the type of value?

File: tests/e2e/performance/test_small_file_workload.py [hunk 21]
```diff
@@ -5,24 +5,316 @@
...
+        Reading all test records from the elasticsearch server into dictionary
+        inside this object
+
+        """
+
+        query = {'query': {'match': {'uuid': self.uuid}}}
+        log.info('Reading all data from ES server')
+        self.all_results = self.es.search(index=self.index,
+                                          body=query,
+                                          size=self.records)
```
Comment (ebenahar) [inline tests/e2e/performance/test_small_file_workload.py:113]:
```suggestion
        self.all_results = self.es.search(
            index=self.index, body=query, size=self.records
        )
```

File: tests/e2e/performance/test_small_file_workload.py [hunk 22]
```diff
@@ -5,24 +5,316 @@
...
 
 log = logging.getLogger(__name__)
 
 
+class SmallFileResultsAnalyse(object):
+    """
+    This class is reading all test results from elasticsearch server (which the
+    ripsaw running of the benchmark is generate), aggregate them by :
+        test operation (e.g. create / delete etc.)
+          sample (for test to be valid it need to run with more the one sample)
```
Comment (fbalak) [inline tests/e2e/performance/test_small_file_workload.py:29]: Indentation on this and next line is not correct.

File: tests/e2e/performance/test_small_file_workload.py [hunk 23]
```diff
@@ -5,24 +5,316 @@
...
+            host (test can be run on more then one pod {called host})
+
+    it generate results for the all test as one unit which will be valid only
```
Comment (fbalak) [inline tests/e2e/performance/test_small_file_workload.py:32]:
```suggestion
    it generates results for all tests as one unit which will be valid only
```

File: tests/e2e/performance/test_small_file_workload.py [hunk 24]
```diff
@@ -5,24 +5,316 @@
...
+        log.info(f'the results data is {self.results}')
+        self.es.index(index=self.new_index, doc_type='_doc', body=self.results)
+
+    def thread_read(self, host, op, snum):
+        """
+        This method read all threads record of one host / operation and sample
+
+        Args:
+            host (str): the name of the pod that ran the test
+            op (str): the operation that tested.
```
Comment (fbalak) [inline tests/e2e/performance/test_small_file_workload.py:130]:
```suggestion
            op (str): the operation that is tested
```

File: tests/e2e/performance/test_small_file_workload.py [hunk 25]
```diff
@@ -5,24 +5,316 @@
...
+                        else:
+                            res[self.managed_keys[key]['name']] = [val]
+                    except Exception:
+                        pass
+        res = self.aggregate_threads_results(res)
+        return res
+
+    def aggregate_threads_results(self, res):
+        """
+        Aggregation one section of results, this can be threads in host,
```
Comment (fbalak) [inline tests/e2e/performance/test_small_file_workload.py:161]:
```suggestion
        Aggregation of one section of results, this can be threads in host,
```

File: tests/e2e/performance/test_small_file_workload.py [hunk 26]
```diff
@@ -125,16 +429,50 @@ def test_smallfile_workload(self, ripsaw, file_size, files, threads, samples, in
...
+                test_status = full_results.aggregate_samples_results()
+                full_results.write()
                 break
 
             if timeout < (time.time() - start_time):
-                raise TimeoutError(f"Timed out waiting for benchmark to complete")
+                raise TimeoutError("Timed out waiting for benchmark to complete")
             time.sleep(30)
-        assert not get_logs_with_errors()
+        assert (not get_logs_with_errors() and test_status), 'Test Failed'
```
Comment (fbalak) [inline tests/e2e/performance/test_small_file_workload.py:478]: Maybe it would be more informative if it was in two asserts:
```
assert not get_logs_with_errors()
assert test_status
```
You would know instantly which one of those conditions failed. Although, I think that in case of a failure there will be printed output of those two variables so it might be ok.

File: tests/e2e/performance/test_small_file_workload.py (hunk 20 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py:97]: As it said in the description : any kind of data type - str/list/dict/int/float

File: tests/e2e/performance/test_small_file_workload.py (hunk 25 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 24 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 23 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 22 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 20 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py:98]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 19 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: done

File: ocs_ci/ocs/ocp.py (hunk 17 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py:849]: done

File: tests/e2e/performance/test_small_file_workload.py (hunk 21 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: fixed

File: ocs_ci/ocs/ocp.py (hunk 18 above)
Comment (Avilir) [reply ocs_ci/ocs/ocp.py]: moved.

File: tests/e2e/performance/test_small_file_workload.py (hunk 26 above)
Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py:480]: I put the together since i want both of them will be check.
if they will be in separate calls and the first one will assert, the second one will not be tested.

Comment (fbalak) [reply tests/e2e/performance/test_small_file_workload.py:480]: Ok, that makes sense.

File: tests/e2e/performance/test_small_file_workload.py [hunk 27]
```diff
@@ -40,40 +42,77 @@ class TestSmallFileWorkload(E2ETest):
...
+        def add_value(res, key, value):
+            if key not in res.keys():
+                res[key] = [value]
+            else:
+                res[key].append(value)
+            return res
+        op = ''
+        for line in logs.split('\n'):
+            log.debug(line)
+            if "operation :" in line:
```
Comment (bengland2) [inline tests/e2e/performance/test_small_file_workload.py]: @Avilir , try smallfile --output-format=json to make your life easier here.    This kind of CI is exactly why I added that feature to it.

Comment (Avilir) [reply tests/e2e/performance/test_small_file_workload.py]: @bengland2 you are looking on outdated file, i delete this function, and instead of taking the results from the log file i pull them directly from the ElastcSearch server.
